APIM Monetization with Moesif

Menuka-Senevirathne opened the pull request:
This introduces the Moesif as the default monetization platform which can be configured with a preferred billing engine such as Stripe, Chargebee etc.

With this approach the Moesif handles both analytics and the metering aspect of the Monetization.This PR includes the basic implementation for handling dynamic commercial tier (Subscription Policy) and generating invoice/billing reports.

Note
Previous efforts on distinguishing analytics and monetization [1] & [2] was referred while doing this implementation.
[1] https://github.com/wso2-extensions/wso2-am-stripe-plugin/pull/61/files#diff-f1453b75623014a64adbdcac085d52d53b3575b0ca3c16341772911b47df4d7f
[2] https://github.com/wso2/carbon-apimgt/pull/13103/files#diff-25d07acd4b3d6313a4822f9c9943c808f53454b09ada31a1ccc3623539fdd710

Menuka-Senevirathne committed:
Initial commit for Moesif based monetization

Menuka-Senevirathne committed:
Create the BillingEngine interface and billing engine specific models

Menuka-Senevirathne committed:
Change the implementation to get the billing usage report from the Moesif instead of the invoice from the Stripe

Menuka-Senevirathne committed:
Move common Moesif method to the utils

Menuka-Senevirathne committed:
Add the license header, remove the parent tag and update the version

Menuka-Senevirathne committed:
Add the license header

Menuka-Senevirathne committed:
Remove BillingProvider.java, unused imports and variables

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/wso2-extensions/wso2-am-stripe-plugin?pullRequest=69) <br/>All committers have signed the CLA.

shilmyhasan commented on pom.xml:
```diff
@@ -1,19 +1,30 @@
 <?xml version="1.0" encoding="UTF-8"?>
+<!--
+        *  Copyright (c) 2005-2011, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
License header need to be upgraded.

Menuka-Senevirathne replied:
Fixed with b55f17a463a72a38dbd70b097bae10fc495c9d39 commit

shilmyhasan commented on src/main/java/org/wso2/apim/monetization/impl/util/MonetizationUtils.java:
```diff
@@ -0,0 +1,535 @@
+
+        try {
+            //get the application key of platform account from tenant conf json file
+            JSONObject tenantConfig = APIUtil.getTenantConfig(tenantDomain);
```
Can we set the Moesif Application key to a static variable and then load the tenant config only if the key is not set yet.

Menuka-Senevirathne replied:
Addressed with  b55f17a463a72a38dbd70b097bae10fc495c9d39 commit

Menuka-Senevirathne committed:
Replace the updated license header

Menuka-Senevirathne committed:
Implement a caching mechanism for the moesifApplicationKey

Introduce moesifApplicationKeyCache and load the key from the tenant configs only if it's unavailable in the cache
